src/signatory/signature_inversion_module.py

```diff
@@ -0,0 +1,115 @@
+
+    """
+
+    batch = signature.shape[0]
+    B = torch.cat(batch * [torch.eye(channels)])
+    new_shape = [batch] + [channels] + [1] * (insertion_position - 1) + [channels] + [1] * (depth + 1 -
+                                                                                            insertion_position)
+    repeat_points = [1, 1] + [channels] * (insertion_position - 1) + [1] + [channels] * (depth + 1 - insertion_position)
+    new_B = B.view(new_shape)
+    new_B = new_B.repeat(repeat_points)
+
+    last_signature_term = smodule.extract_signature_term(signature, channels, depth)
+    last_signature_term = last_signature_term.view([batch] + [channels] * int(depth)).unsqueeze(insertion_position)
+    repeat_points_sig = [1, channels] + [1] * (insertion_position - 1) + [channels] + [1] * (
+                depth + 1 - insertion_position)
+    sig_new_tensor = last_signature_term.unsqueeze(1).repeat(repeat_points_sig)
+
+    A = (new_B * sig_new_tensor).flatten(start_dim=2)
+    return A
+
+
+def solve_optimization_problem(signature, insertion_position, depth, channels):
+    """This function solves the optimization problem that allows to approximate the derivatives of the path.
+
+    Arguments:
+        signature (:class:`torch.Tensor`): As :func:`signatory.invert_signature`.
+
+        insertion_position (int): Insertion spot.
+
+        depth (int): As :func:`signatory.invert_signature`.
+
+        channels (int): As :func:`signatory.invert_signature`.
+
+    Returns:
+        The :class:`torch.Tensor` approximation of the gradient of the streams on the interval indexed by
+        :attr:`insertion_position'. It is of shape :math:`(N, \text{channels})'.
+    """
+
+    A_matrix = get_insertion_matrix(signature[:, :-channels ** depth], insertion_position, depth - 1, channels)
+    b_vector = depth * smodule.extract_signature_term(signature, channels, depth)
+
+    x_optimal = torch.matmul(A_matrix, b_vector.unsqueeze(-1)).squeeze(-1)
+
+    sign_1 = smodule.extract_signature_term(signature, channels, depth - 1)
+
+    return x_optimal / (torch.norm(sign_1, dim=1).unsqueeze(-1) ** 2)
+
+
+def invert_signature(signature: torch.Tensor, depth: int, channels: int,
+                     initial_position: Optional[torch.Tensor] = None) -> torch.Tensor:
+    r"""Invert the signature with the insertion algorithm: reconstruct a stream of data given its signature. Given that
+    the signature is invariant by translation, the initial position of the stream is not recovered.
+
+    The input :attr:`signature` is the signature transform of depth :attr:`depth` of a batch of paths: it should be a
+    result from the :func:`signatory.signature` function. The output is a tensor of shape :math:`(N, L, C)`,
+    where :math:`N` is the batch size, :math:`L` is the length of the reconstructed stream of data, with
+    :math:`L = depth + 1`, and :math:`C` denotes the number of channels.
+
```

Comment:
Can you add a reference to the paper in the docstring?

Reply:
See above, I will add the reference when I put the paper on arXiv.